### ISSUE TITLE
Enhance VariableRequest Struct for Comprehensive Parsing of Grafana Request Payload

### ISSUE THREAD

**Issue by SebiBurchi95** [clambin/grafana-json-server#46]
I've encountered a requirement to accurately parse incoming JSON request payloads from Grafana, which include variable data alongside a specified time range. The current implementation of VariableRequest struct does not fully accommodate the structure of the payloads I am receiving, particularly with respect to capturing additional properties within the payload object.

```
type VariableRequest struct {
	Target VariableTarget `json:"payload"`
	Range  struct {
		From time.Time `json:"from"`
		To   time.Time `json:"to"`
		Raw  struct {
			From string `json:"from"`
			To   string `json:"to"`
		} `json:"raw"`
	} `json:"range"`
}
```

I need to parse the request body in order to take some args. Grafana is sending a request like this:
**{"payload":{"target":"jean","args":"mimi"},"range":{"from":"2024-02-20T15:54:52.955Z","to":"2024-02-20T21:54:52.955Z","raw":{"from":"now-6h","to":"now"}}}**

Can we redefine the VariableRequest struct to more accurately map to the Grafana request payload, ensuring we can extract all necessary information, including any arguments specified within the payload.
Thanks.

**Comment by SebiBurchi95** [comment]
@clambin please have a look here

**Comment by clambin** [comment]
Good idea. I’ll look into it.

**Comment by SebiBurchi95** [comment]
@clambin thank you. let me know if I can help you. I really need this change

**Comment by clambin** [comment]
no worries.  will be out of the country for a week, so will take some time.  In the meantime, how can I reproduce this?  I've only seen GrafanaJsonDatasource produce a VariableTarget with a "target". Not an "args" field.  

Been a while since I worked in this package, so appreciate if you could give me some more info.

**Comment by SebiBurchi95** [comment]
@clambin the idea is that you can manipulate how to send the request for a query variable.
Please check the image below:
<img width="1477" alt="image" src="https://github.com/clambin/grafana-json-server/assets/145971563/a7d3bf98-f4d3-4951-a9a0-635cf54ff34e">

So, I think we can change the implementation to also support a Raw Json.

**Comment by SebiBurchi95** [comment]
@clambin can you help me with an estimate for this? Or, do you want me to make the changes?

**Comment by clambin** [comment]
PR's are always welcome :)

The thing that holds me off from implementing this is that the currently implementation is fundamentally wrong. :)  The implementation of the /variable endpoint assumes there's a "target" field in the requests' "payload" field.  It uses this to route the request to the right VariableFunc (held in the Server.variables map).  

But actually, we can make no assumptions about the layout of the payload: it's a freeform text field.  It's not even guaranteed to have a "target" field.  

grafana-json-server tries to be as non-opinionated as possible. Whatever GrafanaDataSource sends, we should support, withour assumptions or specifical conventions. 

So I guess we should have only /variable handler (no map of targets) with a generic callback and pass the (raw) payload to the user so they can decide what to do with it?  So basically:

- WithVariable(name, handler) becomes WithVariableHandler(handler)
- hander becomes func(json.RawMessage) ([]Variable, error)

A little more work on the user's side, but more in spirit with that grafana-json-server tries to do, I think.

**Comment by SebiBurchi95** [comment]
@clambin if you don't enable "Raw JSON" button for the Grafana Query, you will have a "target" field in the request payload.
Check this out:
<img width="1477" alt="image" src="https://github.com/clambin/grafana-json-server/assets/145971563/d568e1e7-4932-43c7-99fd-7d190a4b3f20">

**Comment by clambin** [comment]
I get that.  But if the user does use "Raw JSON", there's no guarantee that there will be a "target" field.  So I'm leaning towards a single /variable handler that passes the raw JSON to the user, so they can parse it according to how they've set up the query. Something like: 

```
func Example_variable() {
	s := grafanaJSONServer.NewServer(
		grafanaJSONServer.WithVariable(variableFunc),
	)

	_ = http.ListenAndServe(":8080", s)
}

func variableFunc(payload json.RawMessage) ([]grafanaJSONServer.Variable, error) {
	type query struct {
		Target string `json:"target"`
		Args   struct {
			Type string `json:"type"`
		} `json:"args"`
	}

	var req query
	var variables []grafanaJSONServer.Variable
	err := json.Unmarshal(payload, &req)
	switch req.Target {
	case "resources-of-type":
		switch req.Args.Type {
		case "Cluster":
			// determine variables
		}
	}

	return variables, err
}
```

It makes writing a handler a little more complicated, but it doesn't make any assumptions on the query structure. Thoughts?

**Comment by clambin** [comment]
@SebiBurchi95:  if you don't mind trying non-released code, I created a new version on the rawjson branch that should support your use case.

Basically, VariableRequest now contains the Target, as well as the Payload.  In case of Raw JSON, Payload now contains the full JSON object from the Query setting in Grafana's variable definition.

Target still contains the extracted "target" field.  If the object doesn't contain a "target" field, Target will be blank and the user should register a VariableFunc with a blank target.

Was looking for something more intuitive, but I guess this makes for a simple API with relatively few changes from the existing API.

Let me know what you think.

**Comment by clambin** [comment]
fixed in #52 and released in v0.6.0.
